kernel/drivers/LegacyATA.cpp

```diff
@@ -0,0 +1,326 @@
+#define ATA_REG_LBA1 0x04
+#define ATA_REG_LBA2 0x05
+#define ATA_REG_HDDEVSEL 0x06
+#define ATA_REG_COMMAND 0x07
+#define ATA_REG_STATUS 0x07
+#define ATA_REG_SECCOUNT1 0x08
+#define ATA_REG_LBA3 0x09
+#define ATA_REG_LBA4 0x0A
+#define ATA_REG_LBA5 0x0B
+#define ATA_REG_CONTROL 0x0C
+#define ATA_REG_ALTSTATUS 0x0C
+#define ATA_REG_DEVADDRESS 0x0D
+
+// IO Ports
+#define ATA_PRIMARY_IO 0x1F0
+#define ATA_SECONDARY_IO 0x170
+
+// LBA modes
+#define ATA_28LBA_MAX 0x0FFFFFFF
+#define ATA_48LBA_MAX 0xFFFFFFFFFFFF
+#define ATA_SECTOR_SIZE 512
+
+LegacyATA::LegacyATA(DeviceAddress address) : LegacyDevice(address, DeviceClass::DISK)
+{
+    switch (address.legacy())
+    {
+    case LEGACY_ATA0:
+        _bus = ATA_PRIMARY;
+        _drive = ATA_MASTER;
+        break;
+    case LEGACY_ATA1:
+        _bus = ATA_PRIMARY;
+        _drive = ATA_SLAVE;
+        break;
+    case LEGACY_ATA2:
+        _bus = ATA_SECONDARY;
+        _drive = ATA_MASTER;
+        break;
+    case LEGACY_ATA3:
+        _bus = ATA_SECONDARY;
+        _drive = ATA_SLAVE;
+        break;
+
+    default:
+        break;
+    }
+
+    identify();
+}
+
+void LegacyATA::select()
+{
+    const uint16_t io_port = _bus == ATA_PRIMARY ? ATA_PRIMARY_IO : ATA_SECONDARY_IO;
+    out8(io_port + ATA_REG_HDDEVSEL, _drive == ATA_MASTER ? 0xA0 : 0xB0);
+}
+
+size_t LegacyATA::size(FsHandle &handle)
+{
+    __unused(handle);
+    return _num_blocks * ATA_SECTOR_SIZE;
+}
+
+void LegacyATA::identify()
+{
+    select();
+    const uint16_t io_port = _bus == ATA_PRIMARY ? ATA_PRIMARY_IO : ATA_SECONDARY_IO;
+
+    /* ATA specs say these values must be zero before sending IDENTIFY */
+    out8(io_port + ATA_REG_SECCOUNT0, 0);
+    out8(io_port + ATA_REG_LBA0, 0);
+    out8(io_port + ATA_REG_LBA1, 0);
+    out8(io_port + ATA_REG_LBA2, 0);
+
+    /* Now, send IDENTIFY */
+    out8(io_port + ATA_REG_COMMAND, ATA_CMD_IDENTIFY);
+    logger_info("Sent IDENTIFY");
+
+    /* Now, read status port */
+    uint8_t status = in8(io_port + ATA_REG_STATUS);
+    if (status)
+    {
+        /* Now, poll untill BSY is clear. */
+        while ((in8(io_port + ATA_REG_STATUS) & ATA_SR_BSY) != 0)
+            ;
+
+        do
+        {
+            status = in8(io_port + ATA_REG_STATUS);
+
+            if (status & ATA_SR_ERR)
+            {
+                logger_error("%s%s has ERR set. Disabled.", _bus == ATA_PRIMARY ? "Primary" : "Secondary",
+                             _drive == ATA_PRIMARY ? " master" : " slave");
+
+                return;
+            }
+        } while (!(status & ATA_SR_DRQ));
+
+        logger_info("%s%s is online.", _bus == ATA_PRIMARY ? "Primary" : "Secondary", _drive == ATA_PRIMARY ? " master" : " slave");
+
+        // Read back the data
+        for (int i = 0; i < 256; i++)
+        {
+            _ide_buffer[i] = in16(io_port + ATA_REG_DATA);
+        }
+
+        _exists = true;
+
+        // Parse some infos from the identify buffer
+        char *model_buf = (char *)malloc(40);
```

Comment:
This buffer can be allocated on the stack

Reply:
also new and delete is available in the kernel ;)

Reply:
Okay, ideally i wanted to allocate the string ("_model") with a size and directly write into it. However i was lacking too many operators in **String**. Will fix this